Fix/metatada issues

josefinalliende opened the pull request:
## Description
As part of the effort to remove the metadata cache provider (#541), this PR removes more references and introduces fixes along the way. One fix is that the edit profile screen now updates metadata correctly (#558), but the main fix is that metadata of active account is now updated. Fixes also #572 

With these changes, the metadata of the active account is updated via the polling provider (which invalidates the active account provider). This ensures that metadata updates from other clients are reflected in Whitenoise.  



**Commit by commit:**
- Removed remaining unnecessary metadata cache provider code.  
- Added a simple `userProfileDataProvider` that fetches account metadata directly from the DB.  
- Replaced metadata cache provider with `userProfileDataProvider` in multiple places.  
- Updated polling provider to invalidate the active account provider.  
- Simplified several usages of the active account provider by switching to the active pubkey provider where possible.  
- Fixed a new “blinking” issue in the chat list screen by moving the active account provider watch into a dedicated widget (avoids rebuilding the entire screen).  
- Added changes to the changelog.  
- Fixes active account provider to use get default blossom url method and changes local default to port 3000 to match the port run by default in the whitenoise rust repo.
- Changes pubkey conversion methods to sync, adding support for ContactModel.fromMetadata to receive pubkey in hex ur npub format. This way it fixes a problem commented in #562  with npubs showing as hex

Short video of how now when updating metadata in primal it updates in WN


https://github.com/user-attachments/assets/d68dd67b-9d6c-4766-be6f-a1f178466ba0


And the other way, changing profile picture in WN and seeing updates in primal


https://github.com/user-attachments/assets/0d13e8e6-7d21-455a-99b4-f33383b4534f




## Type of Change

<!--- Put an `x` in all the boxes that apply: -->

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore
- [ ] 🧪 Tests

## Checklist

<!-- Please make sure you've done the following before submitting your PR: -->

- [x] Run `just precommit` to ensure that formatting and linting are correct
- [ ] Run `just check-flutter-coverage` to ensure that flutter coverage rules are passing
- [x] Updated the `CHANGELOG.md` file with your changes (if they affect the user experience)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * New user profile data provider and header avatar component for account display.

* **Bug Fixes**
  * Fixes profile editing when metadata is outdated; more reliable profile loading across chats, groups, welcomes and follows.

* **Refactor**
  * Migration to pubkey-based flows; metadata cache and related debug utilities removed.

* **UI**
  * Invite/welcome and new-chat screens now use unified profile data; avatar/header updates.

* **Developer**
  * Removed Clear Cache action; debug upload server URL adjusted.

* **Tests**
  * Added unit tests for the user profile data provider.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

codeswot commented on lib/config/providers/group_provider.dart:
```diff
@@ -1061,17 +1048,6 @@ extension GroupMemberUtils on GroupsNotifier {
     return otherMembers.first;
   }
 
-  Future<ContactModel?> getFirstOtherMember(String? groupId, String? currentUserNpub) async {
```
is this no longer being used  ?

josefinalliende replied:
yes, it wasn't used anywhere

josefinalliende commented on lib/config/providers/user_profile_data_provider.dart:
```diff
@@ -0,0 +1,40 @@
+import 'package:whitenoise/src/rust/api/users.dart' as wn_users_api;
+import 'package:whitenoise/src/rust/api/users.dart' show User;
+
+abstract class WnUsersApi {
```
If I call the rust method directly inside the provider, it becomes very hard to test. The idea behind adding dependency injection is to make it easier to write tests with mocks. I’ll try a different approach to see if I can achieve the same with a const, while still keeping the provider testable.

josefinalliende replied:
So you were right about the abstract class, it wasn't necessary. Now I made changes so that the abstract class is removed but still has dependency injection to make it easier to test

codeswot replied:
> If I call the rust method directly inside the provider, it becomes very hard to test. The idea behind adding dependency injection is to make it easier to write tests with mocks. I’ll try a different approach to see if I can achieve the same with a const, while still keeping the provider testable.

I love that, this was one of the reasons i proposed having a datasource to handle api calls.
we can come back to this later and do it in a clean unified way.

josefinalliende commented on lib/config/providers/chat_provider.dart:
```diff
@@ -65,9 +64,8 @@ class ChatNotifier extends Notifier<ChatState> {
-      final activeAccountState = await ref.read(activeAccountProvider.future);
-      final activeAccount = activeAccountState.account;
-      if (activeAccount == null) {
+      final activePubkey = ref.read(activePubkeyProvider);
```
I'll check that out. @codeswot  proposed to use .nullOrEmpty and I liked that one, I think it will look prettier

codeswot replied:
> Related to null promotion, I ended up choosing your suggestion. The .nullOrEmpty required to add it to string extension file but for String? instead of the existing ones for String and also gave me warnings after cause the compiler god confused thinking that I was passing a String? to an argument that could only be a String. Null promotion works better in this case I think, so thanks ❤️

You mean when using activePubkey string in the npubFromHexPubkey() method, it complains that it could be null.
I agree Null promotion works better here as it keeps the context of the variable being checked if null

josefinalliende commented on lib/config/providers/nostr_keys_provider.dart:
```diff
@@ -42,11 +40,11 @@ class NostrKeysNotifier extends Notifier<NostrKeysState> {
       // Load npub and nsec directly from hex pubkey string
-      final npubString = await npubFromHexPubkey(hexPubkey: activeAccount.pubkey);
-      final nsecString = await exportAccountNsec(pubkey: activeAccount.pubkey);
+      final npubString = await npubFromHexPubkey(hexPubkey: activePubkey);
```
I think the variable name is a bit misleading and I can change it,  npubFromHexPubkey is actually converting from hex format to bech32. I’ll double-check if this conversion is really needed.

josefinalliende commented on lib/config/providers/user_profile_data_provider.dart:
```diff
@@ -0,0 +1,40 @@
+import 'package:whitenoise/domain/models/contact_model.dart';
+import 'package:whitenoise/src/rust/api/users.dart' as wn_users_api;
+import 'package:whitenoise/src/rust/api/users.dart' show User;
+
```
I added that but in the exptension of pubkey validations

josefinalliende commented on lib/config/providers/welcomes_provider.dart:
```diff
@@ -316,13 +311,12 @@ class WelcomesNotifier extends Notifier<WelcomesState> {
-      final newWelcomes = await pendingWelcomes(pubkey: activeAccount.pubkey);
+      final newWelcomes = await pendingWelcomes(pubkey: activePubkey);
 
       final currentWelcomes = state.welcomes ?? [];
```
But this PR did not changed welcomes. Can it be a different issue?

josefinalliende replied:
Oh ok, thanks for the explanation!

josefinalliende commented:
Oh no, tests are failing 😢 ... will fix them tomorrow then 😴

josefinalliende commented:
> @codeswot @Quwaysim Ithink I’ve addressed almost all of your comments, along with a few from CodeRabbit. I added support to the user profile data provider for handling both hex and bech32 pubkeys. This required changes in the ContactModel, public key extension methods, and some widgets — since the ContactModel now needs to be built with await (because it uses the Rust crate to transform hex into npub). This also fixes the issue that in some widgets npub was being shown in hex format instead of npub format.

By making the pubkey conversion methods synchronous, I no longer need to update widgets with FutureBuilder to build contact models. The issue with npubs showing as hex is still fixed, but now in a much simpler way 😄. I also updated the change list in the PR description to make it easier to re-review

josefinalliende commented on lib/config/providers/nostr_keys_provider.dart:
```diff
@@ -42,11 +40,11 @@ class NostrKeysNotifier extends Notifier<NostrKeysState> {
-      final nsecString = await exportAccountNsec(pubkey: activeAccount.pubkey);
+      final npubString = await npubFromHexPubkey(hexPubkey: activePubkey);
+      final nsecString = await exportAccountNsec(pubkey: activePubkey);
 
```
names where improved here.

josefinalliende commented on rust/src/api/utils.rs:
```diff
@@ -9,27 +9,17 @@ use nostr_mls::prelude::GroupId;
 use nostr_sdk::prelude::*;
 pub use whitenoise::Whitenoise;
 
-#[frb]
```
It was an error to delete this method

josefinalliende commented on rust/src/api/utils.rs:
```diff
@@ -9,27 +9,17 @@ use nostr_mls::prelude::GroupId;
     Ok(pubkey.to_hex())
 }
 
-#[frb]
```
Same for this one, I'm biringing them back

josefinalliende commented on lib/config/providers/metadata_cache_provider.dart:
```diff
@@ -110,9 +109,8 @@ class MetadataCacheNotifier extends Notifier<MetadataCacheState> {
-      final activeAccountState = await ref.read(activeAccountProvider.future);
-      final activeAccount = activeAccountState.account;
-      if (activeAccount == null) {
+      final activePubkey = ref.read(activePubkeyProvider) ?? '';
+      if (activePubkey.isEmpty) {
         throw StateError('No active account found');
       }
```
This provider is going to be removed very very soon. So it's not worth improving this part of the code

josefinalliende commented on lib/config/providers/active_account_provider.dart:
```diff
@@ -137,10 +137,10 @@ Future<FlutterMetadata> _fetchMetadata(WnAccountsApi accountsApi, String pubkey)
   Future<ActiveAccountState> build() async {
-    final activePubkey = ref.watch(activePubkeyProvider);
+    final activePubkey = ref.watch(activePubkeyProvider) ?? '';
     final accountsApi = ref.read(wnAccountsApiProvider);
 
-    if (activePubkey == null || activePubkey.isEmpty) {
+    if (activePubkey.isEmpty) {
       _logger.fine('No active pubkey set');
       return const ActiveAccountState();
     }
```
In #597  I'm handling it in a better way bu also checking formats before saving and reading

josefinalliende commented on lib/domain/models/contact_model.dart:
```diff
@@ -33,32 +33,11 @@ class ContactModel {
-    final nip05 = _sanitizeString(metadata.nip05);
-    final lud16 = _sanitizeString(metadata.lud16);
-    final picture = _sanitizeUrl(metadata.picture);
-
-    return ContactModel(
-      displayName: displayName.isNotEmpty ? displayName : 'Unknown User',
-      publicKey: user.pubkey,
+      publicKey: npub,
       imagePath: picture,
```
This is also fixed in #597

josefinalliende commented on lib/src/rust/api/utils.dart:
```diff
@@ -9,18 +9,12 @@ import '../frb_generated.dart';
-
-Future<String> npubFromHexPubkey({required String hexPubkey}) =>
+String npubFromHexPubkey({required String hexPubkey}) =>
     RustLib.instance.api.crateApiUtilsNpubFromHexPubkey(hexPubkey: hexPubkey);
 
-Future<String> hexPubkeyFromNpub({required String npub}) =>
+String hexPubkeyFromNpub({required String npub}) =>
     RustLib.instance.api.crateApiUtilsHexPubkeyFromNpub(npub: npub);
```
This is now fixed in #597 too

josefinalliende commented on lib/ui/chat/widgets/chat_header_widget.dart:
```diff
@@ -51,19 +51,19 @@ class GroupChatHeader extends ConsumerStatefulWidget {
   void didUpdateWidget(GroupChatHeader oldWidget) {
     super.didUpdateWidget(oldWidget);
     if (oldWidget.group.nostrGroupId != widget.group.nostrGroupId) {
-      _groupNpubFuture = npubFromHexPubkey(hexPubkey: widget.group.nostrGroupId);
+      _groupNpub = npubFromHexPubkey(hexPubkey: widget.group.nostrGroupId);
     }
```
this is fixed with pubkey formatter in #597 too

josefinalliende commented on lib/ui/settings/profile/switch_profile_bottom_sheet.dart:
```diff
@@ -67,7 +67,7 @@ class _SwitchProfileBottomSheetState extends ConsumerState<SwitchProfileBottomSh
       try {
         if (originalPubKey.startsWith('npub1')) {
-          final hex = await hexPubkeyFromNpub(npub: originalPubKey);
+          final hex = hexPubkeyFromNpub(npub: originalPubKey);
           _pubkeyToHex[originalPubKey] = hex;
```
now it's fixed in #597 by using pubkey formatter

josefinalliende commented on lib/ui/chat/widgets/chat_header_widget.dart:
```diff
@@ -51,19 +51,19 @@ class GroupChatHeader extends ConsumerStatefulWidget {
   void initState() {
     super.initState();
-    _groupNpubFuture = npubFromHexPubkey(hexPubkey: widget.group.nostrGroupId);
+    _groupNpub = npubFromHexPubkey(hexPubkey: widget.group.nostrGroupId);
   }
```
fixed in #587

josefinalliende commented on lib/domain/services/dm_chat_service.dart:
```diff
@@ -24,8 +25,16 @@ class DMChatService {
 
       if (otherMember != null) {
-        final metadataCacheNotifier = ref.read(metadataCacheProvider.notifier);
-        final contactModel = await metadataCacheNotifier.getContactModel(otherMember.publicKey);
+        final user = await wn_users_api.getUser(pubkey: otherMember.publicKey);
+        final otherMemberPubkey = otherMember.publicKey;
+        String otherMemberNpubPubkey = otherMemberPubkey;
+        if (otherMemberPubkey.isValidHexPublicKey) {
+          otherMemberNpubPubkey = npubFromHexPubkey(hexPubkey: otherMemberPubkey);
+        }
+        final contactModel = ContactModel.fromMetadata(
+          pubkey: otherMemberNpubPubkey,
+          metadata: user.metadata,
+        );
```
fixes, already transformed to hex in #597
